Framework/Kernel/test/CowPtrTest.h

```diff
@@ -0,0 +1,66 @@
+#ifndef COW_PTR_TEST_H_
+#define COW_PTR_TEST_H_
+
+#include <cxxtest/TestSuite.h>
+#include "MantidKernel/cow_ptr.h"
+#include <boost/shared_ptr.hpp>
+#include <boost/make_shared.hpp>
+
+using namespace Mantid::Kernel;
+
+namespace {
+
+struct MyType {
+public:
+  MyType() : value(0) {}
+  MyType(int val) : value(val) {}
+  int value;
+};
+}
+
+class CowPtrTest : public CxxTest::TestSuite {
+public:
+  void testDefaultConstruct() {
+
+    cow_ptr<MyType> cow{};
+    TSM_ASSERT_EQUALS("Should give us the default of the default constructed T",
+                      0, cow->value);
+  }
+
+  void testConstructorByPtr() {
+
+    MyType *nullMyType = nullptr;
+    TSM_ASSERT_THROWS("Source cannot be null", cow_ptr<MyType>{nullMyType},
+                      std::invalid_argument &);
+
+    auto *resource = new MyType{2};
+    cow_ptr<MyType> cow{resource};
+
+    TSM_ASSERT_EQUALS("COW does not hold the expected value", 2, cow->value);
+  }
+
+  void testConstructorBySptr() {
+
+    int value = 3;
+    auto resource = boost::make_shared<MyType>(value);
+    cow_ptr<MyType> cow{std::move(resource)};
+
+    TS_ASSERT_EQUALS(cow->value, value);
+    TSM_ASSERT("Resource should have been moved", resource.get() == nullptr);
+  }
+
+  void testCow() {
+
+    int value = 3;
+    cow_ptr<MyType> original{boost::make_shared<MyType>(value)};
+
+    MyType copy = original.access();
```

Comment:
I may be mistaken, but I think this test does not do what you want it to do:

1. Create: `cow_ptr<MyType> original{boost::make_shared<MyType>(value)};`
2. Access: `original.access()`. Reference count is 1, so no copy is made, I think.
3. Assignment: `MyType copy = original.access();`. `access()` returns a reference, you assign to value, so you are invoking the copy constructor of `MyType`.

That is, I think the `ASSERT` for the address below works because you copy the type. I think what we really want is a `cow_ptr` with reference count > 1, and then access and capture the result be reference.